src/module_3/evaluation_functions.py

```diff
@@ -0,0 +1,179 @@
+
+import pandas as pd
+import numpy as np
+from sklearn.model_selection import train_test_split
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import  StandardScaler
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import GridSearchCV
+from sklearn.compose import ColumnTransformer
+import category_encoders as ce
+from sklearn.model_selection import StratifiedKFold
+import matplotlib.pyplot as plt
+from datetime import timedelta
+from sklearn.model_selection import PredefinedSplit
+import pyarrow.parquet as pq
+from sklearn import metrics
+from sklearn.metrics import roc_auc_score, confusion_matrix
+
+
+def plot_precision_recall_vs_threshold(clf, X, y):
+    if hasattr(clf, 'predict_proba'):
+        model_probs = clf.predict_proba(X)
+        model_probs = model_probs[:, 1]
+    else:
+        model_probs = clf.decision_function(X)
+
+    model_precision, model_recall, thresholds = metrics.precision_recall_curve(y, model_probs)
+    thresholds = np.insert(thresholds, thresholds.size, 1., axis=0)
+
+    plt.figure()
+    plt.plot(thresholds, model_precision, 'r', label='Precision')
+    plt.plot(thresholds, model_recall, 'b', label='Recall')
+    plt.xlabel('Threshold Value')
+    plt.title('Precision and Recall vs Threshold')
+    plt.legend(loc='center left', bbox_to_anchor=(1, 0.5))
+    plt.show()
+
+def threshold_classification(positive_class_probs, threshold=0.5):
+    return (positive_class_probs >= threshold).astype('int')
+
+def find_best_threshold(model, X_train, y_train, X_val, y_val, metric='fscore'):
+    if hasattr(model, 'predict_proba'):
+        model_probs = model.predict_proba(X_train)
+        model_probs = model_probs[:, 1]
+        model_probs_val = model.predict_proba(X_val)
+        model_probs_val = model_probs_val[:, 1]
+    else:
+        model_probs = model.decision_function(X_train)
+        model_probs_val = model.decision_function(X_val)
+        
+    if (metric == 'fscore'):
+        model_precision, model_recall, thresholds = metrics.precision_recall_curve(y_train, model_probs)  
+        performance_thresholds = [metrics.f1_score(y_val, threshold_classification(model_probs_val, t)) * 100.0 for t in thresholds]
+        best_threshold_index = np.argmax(performance_thresholds)
+    elif (metric == 'auc'):
+        model_fpr, model_tpr, thresholds = metrics.roc_curve(y_train, model_probs)
+        performance_thresholds = [metrics.roc_auc_score(y_val, threshold_classification(model_probs_val, t)) * 100.0 for t in thresholds]
+        best_threshold_index = np.argmax(performance_thresholds)
```

Comment:
Firstly, I think this is wrongly calculated. `metrics.roc_auc_score` should receive `proba` as the second argument.

Secondly, can we numerically compute what't the optimal threshold without considering the business problem? As we discussed before, the point closest to the top left corner doesn't need to be the optimal point for any specific problem. The working point is completely dependant of the problem.